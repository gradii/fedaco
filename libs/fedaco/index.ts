/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license
 */

export * from './src/interface/connection-resolver-interface';
export * from './src/base-grammar';
export * from './src/define/collection';
export * from './src/connection/sql-server-connection';
export * from './src/connection/mysql-connection';
export * from './src/connection/sqlite-connection';
export * from './src/connection/postgres-connection';
export * from './src/fedaco/model-helper-global';
export * from './src/fedaco/fedaco-builder';
export * from './src/fedaco/fedaco-event-manage';
export * from './src/fedaco/model-meta-data';
export * from './src/fedaco/base-model';
export * from './src/fedaco/scope';
export * from './src/fedaco/model-resolver';
export * from './src/fedaco/relations/pivot';
export * from './src/fedaco/relations/pivot-helper';
export * from './src/fedaco/relations/has-one-or-many';
export * from './src/fedaco/relations/has-many-through';
export * from './src/fedaco/relations/morph-many';
export * from './src/fedaco/relations/has-many';
export * from './src/fedaco/relations/belongs-to-many';
export * from './src/fedaco/relations/morph-one-or-many';
export * from './src/fedaco/relations/morph-one';
export * from './src/fedaco/relations/has-one';
export * from './src/fedaco/relations/belongs-to';
export * from './src/fedaco/relations/has-one-through';
export * from './src/fedaco/relations/morph-pivot';
export * from './src/fedaco/relations/morph-to-many';
export * from './src/fedaco/relations/relation';
export * from './src/fedaco/relations/concerns/interacts-with-pivot-table';
export * from './src/fedaco/relations/concerns/can-be-one-of-many';
export * from './src/fedaco/relations/concerns/compares-related-models';
export * from './src/fedaco/relations/concerns/supports-default-models';
export * from './src/fedaco/relations/concerns/as-pivot';
export * from './src/fedaco/relations/concerns/interacts-with-dictionary';
export * from './src/fedaco/relations/morph-to';
export * from './src/fedaco/mixins/soft-deletes';
export * from './src/fedaco/mixins/has-global-scopes';
export * from './src/fedaco/mixins/guards-attributes';
export * from './src/fedaco/mixins/hides-attributes';
export * from './src/fedaco/mixins/has-timestamps';
export * from './src/fedaco/mixins/queries-relationships';
export * from './src/fedaco/mixins/has-relationships';
export * from './src/fedaco/mixins/forward-call-to-query-builder';
export * from './src/fedaco/mixins/has-events';
export * from './src/fedaco/mixins/has-attributes';
export * from './src/fedaco/model-helper';
export * from './src/fedaco/model';
export * from './src/fedaco/fedaco-manage';
export * from './src/fedaco/resolve-connection';
export * from './src/fedaco/encrypter';
export * from './src/fedaco/scopes/soft-deleting-scope';
// export * from './src/dbal/column';
// export * from './src/dbal/table';
// export * from './src/dbal/table-diff';
export * from './src/annotation/column';
export * from './src/annotation/additional-processing';
// export * from './src/annotation/scope';
export * from './src/annotation/enum-relation';
export * from './src/annotation/column/column';
export * from './src/annotation/column/binary.column';
export * from './src/annotation/column/updated-at.column';
export * from './src/annotation/column/primary-generated.column';
export * from './src/annotation/column/integer.column';
export * from './src/annotation/column/array.column';
export * from './src/annotation/column/float.column';
export * from './src/annotation/column/datetime.column';
export * from './src/annotation/column/decimal.column';
export * from './src/annotation/column/created-at.column';
export * from './src/annotation/column/timestamp.column';
export * from './src/annotation/column/primary.column';
export * from './src/annotation/column/object.column';
export * from './src/annotation/column/deleted-at.column';
export * from './src/annotation/column/date.column';
export * from './src/annotation/column/text.column';
export * from './src/annotation/column/boolean.column';
export * from './src/annotation/column/json.column';
export * from './src/annotation/column/currency.column';
export * from './src/annotation/annotation.interface';
export * from './src/annotation/table/table';
export * from './src/annotation/boot';
export * from './src/annotation/relation-using.relation';
export * from './src/annotation/featured-column/auto-number.column';
// export * from './src/annotation/featured-column/paragraph.column';
export * from './src/annotation/relation-column';
export * from './src/annotation/relation-column/has-many-through.relation-column';
export * from './src/annotation/relation-column/morph-one.relation-column';
export * from './src/annotation/relation-column/morph-many.relation-column';
export * from './src/annotation/relation-column/morph-to-many.relation-column';
export * from './src/annotation/relation-column/belongs-to.relation-column';
export * from './src/annotation/relation-column/belongs-to-many.relation-column';
export * from './src/annotation/relation-column/has-many.relation-column';
export * from './src/annotation/relation-column/morphed-by-many.relation-column';
export * from './src/annotation/relation-column/morph-eager-to.relation-column';
export * from './src/annotation/relation-column/has-one.relation-column';
export * from './src/annotation/relation-column/morph-to.relation-column';
export * from './src/annotation/relation-column/has-one-through.relation-column';
export * from './src/connection';
export * from './src/schema/foreign-id-column-definition';
export * from './src/schema/postgres-schema-state';
export * from './src/schema/grammar/mysql-schema-grammar';
export * from './src/schema/grammar/postgres-schema-grammar';
export * from './src/schema/grammar/sqlite-schema-grammar';
export * from './src/schema/grammar/sql-server-schema-grammar';
// export * from './src/schema/grammar/rename-column';
export * from './src/schema/grammar/schema-grammar';
// export * from './src/schema/grammar/change-column';
export * from './src/schema/schema-state';
export * from './src/schema/sqlite-schema-state';
export * from './src/schema/column-definition';
export * from './src/schema/schema-builder';
export * from './src/schema/mysql-schema-state';
export * from './src/schema/foreign-key-definition';
export * from './src/schema/blueprint';
export * from './src/schema/builder/sql-server-schema-builder';
export * from './src/schema/builder/sqlite-schema-builder';
export * from './src/schema/builder/postgres-schema-builder';
export * from './src/schema/builder/mysql-schema-builder';
export * from './src/connector/sql-server-connector';
// export * from './src/connector/mariadb/mariadb-connector';
// export * from './src/connector/mariadb/mariadb-wrapped-connection';
// export * from './src/connector/mariadb/mariadb-wrapped-stmt';
export * from './src/connector/wrapped-connection';
export * from './src/connector/connector';
export * from './src/connector/sqlite/sqlite-wrapped-connection';
export * from './src/connector/sqlite/sqlite-connector';
export * from './src/connector/sqlite/sqlite-wrapped-stmt';
export * from './src/connector/mysql/mysql-wrapped-connection';
export * from './src/connector/mysql/mysql-connector';
export * from './src/connector/mysql/mysql-wrapped-stmt';
export * from './src/connector/wrapped-stmt';
export * from './src/connector/postgres-connector';
export * from './src/connector/connector-interface';
export * from './src/connector/connection-factory';
export * from './src/database-manager';
export * from './src/events/query-executed';
export * from './src/events/connection-event';
export * from './src/events/schema-loaded';
export * from './src/events/transaction-rolled-back';
export * from './src/events/schema-dumped';
export * from './src/events/statement-prepared';
export * from './src/events/no-pending-migrations';
export * from './src/events/transaction-committed';
export * from './src/events/models-pruned';
export * from './src/events/transaction-beginning';
export * from './src/query/sql-node';
export * from './src/query/expr/functions/function-node';
export * from './src/query/expr/functions/abs-function';
export * from './src/query/parser/sql-parser';
export * from './src/query/parser/helper';
// export * from './src/query/parser/_sql-parser-ast';
export * from './src/query/parser/sql-lexer';
export * from './src/query/parser/keywords';
export * from './src/query/parser/ascii-chars';
export * from './src/query/sql-visitor';
export * from './src/query/ast/table-name';
export * from './src/query/ast/path-expression';
export * from './src/query/ast/join-expression';
export * from './src/query/ast/select-insert-source';
export * from './src/query/ast/identifier';
export * from './src/query/ast/column-reference-expression';
export * from './src/query/ast/order-by-element';
export * from './src/query/ast/select-clause';
export * from './src/query/ast/join-on-expression';
export * from './src/query/ast/having-clause';
export * from './src/query/ast/delete-specification';
export * from './src/query/ast/join-clause';
export * from './src/query/ast/index-by';
export * from './src/query/ast/group-by-clause';
export * from './src/query/ast/range-variable-declaration';
export * from './src/query/ast/offset-clause';
export * from './src/query/ast/binary-query-expression';
export * from './src/query/ast/values-insert-source';
export * from './src/query/ast/where-clause';
export * from './src/query/ast/binary-union-query-expression';
export * from './src/query/ast/assignment-set-clause';
export * from './src/query/ast/query-expression';
export * from './src/query/ast/fragment/order/reject-order-element-expression';
export * from './src/query/ast/fragment/union-fragment';
export * from './src/query/ast/fragment/aggregate-fragment';
export * from './src/query/ast/fragment/nested-expression';
export * from './src/query/ast/fragment/raw-fragment';
export * from './src/query/ast/fragment/expression/nested-predicate-expression';
export * from './src/query/ast/fragment/json-path-column';
export * from './src/query/ast/fragment/join-fragment';
export * from './src/query/ast/joined-table';
export * from './src/query/ast/select-scalar-expression';
export * from './src/query/ast/set-clause';
export * from './src/query/ast/update-specification';
export * from './src/query/ast/table-reference-expression';
export * from './src/query/ast/json-path-expression';
export * from './src/query/ast/node-part';
export * from './src/query/ast/from-clause';
export * from './src/query/ast/insert-specification';
export * from './src/query/ast/expression/number-literal-expression';
export * from './src/query/ast/expression/raw-binding-expression';
export * from './src/query/ast/expression/expression';
export * from './src/query/ast/expression/binary-expression';
export * from './src/query/ast/expression/condition-factor-expression';
export * from './src/query/ast/expression/common-value-expression';
export * from './src/query/ast/expression/condition-expression';
export * from './src/query/ast/expression/null-predicate-expression';
export * from './src/query/ast/expression/string-literal-expression';
export * from './src/query/ast/expression/raw-expression';
export * from './src/query/ast/expression/exists-predicate-expression';
export * from './src/query/ast/expression/as-expression';
export * from './src/query/ast/expression/function-call-expression';
export * from './src/query/ast/expression/parenthesized-expression';
export * from './src/query/ast/expression/in-predicate-expression';
export * from './src/query/ast/expression/between-predicate-expression';
export * from './src/query/ast/expression/comparison-predicate-expression';
export * from './src/query/ast/expression/condition-term-expression';
export * from './src/query/ast/order-by-clause';
export * from './src/query/ast/from-table';
export * from './src/query/ast/lock-clause';
export * from './src/query/ast/query-specification';
export * from './src/query/ast/identify-variable-declaration';
export * from './src/query/ast/binding-variable';
export * from './src/query/ast/limit-clause';
export * from './src/helper/pluralize';
export * from './src/helper/fn';
export * from './src/helper/obj';
export * from './src/helper/arr';
export * from './src/helper/str';
export * from './src/helper/constructor';
// export * from './src/helper/_pluralize-values';
export * from './src/query-builder/query-builder-resolver-visitor';
export * from './src/query-builder/grammar/query-grammar';
// export * from './src/query-builder/grammar/oracle-query-grammar';
export * from './src/query-builder/grammar/mysql-query-grammar';
export * from './src/query-builder/grammar/postgres-query-grammar';
export * from './src/query-builder/grammar/sqlite-query-grammar';
export * from './src/query-builder/grammar/sqlserver-query-grammar';
export * from './src/query-builder/ast-factory';
export * from './src/query-builder/mixins/union';
export * from './src/query-builder/mixins/having';
export * from './src/query-builder/mixins/where-predicate';
export * from './src/query-builder/mixins/order-by';
export * from './src/query-builder/mixins/where-common';
export * from './src/query-builder/mixins/build-query';
export * from './src/query-builder/mixins/join';
export * from './src/query-builder/mixins/where-date';
export * from './src/query-builder/mixins/group-by';
export * from './src/query-builder/mixins/limit-offset';
export * from './src/query-builder/mixins/aggregate';
export * from './src/query-builder/processor/mysql-processor';
export * from './src/query-builder/processor/postgres-processor';
export * from './src/query-builder/processor/sqlite-processor';
export * from './src/query-builder/processor/sql-server-processor';
export * from './src/query-builder/builder';
export * from './src/query-builder/connection-interface';
export * from './src/query-builder/grammar.interface';
export * from './src/query-builder/processor-interface';
export * from './src/query-builder/processor';
export * from './src/query-builder/visitor/mysql-query-builder-visitor';
export * from './src/query-builder/visitor/sqlserver-query-builder-visitor';
export * from './src/query-builder/visitor/postgres-query-builder-visitor';
export * from './src/query-builder/visitor/query-builder-visitor';
export * from './src/query-builder/visitor/sqlite-query-builder-visitor';
export * from './src/query-builder/forward-ref';
export * from './src/query-builder/query-builder';
export * from './src/query-builder/ast-helper';
export * from './src/database-config';
export * from './src/database-manager';
export * from './src/database-transaction-record';
export * from './src/database-transactions-manager';
export * from './src/manages-transactions';
export * from './src/query-exception';
