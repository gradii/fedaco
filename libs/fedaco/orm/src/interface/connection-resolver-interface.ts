export interface ConnectionResolverInterface {
  /*Get a database connection instance.*/
  connection(name: string | null);
  /*Get the default connection name.*/
  getDefaultConnection();
  /*Set the default connection name.*/
  setDefaultConnection(name: string);
}
