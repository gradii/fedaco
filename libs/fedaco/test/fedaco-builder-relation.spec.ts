/* tslint:disable:max-line-length */
import { DatabaseTransactionsManager } from '../src/database-transactions-manager';
import { FedacoBuilder } from '../src/fedaco/fedaco-builder';
import { Model } from '../src/fedaco/model';
import { ConnectionResolverInterface } from '../src/interface/connection-resolver-interface';
import { ConnectionInterface } from '../src/query-builder/connection-interface';
import { MysqlQueryGrammar } from '../src/query-builder/grammar/mysql-query-grammar';
import { Processor } from '../src/query-builder/processor';
import { JoinClauseBuilder, QueryBuilder } from '../src/query-builder/query-builder';
import { SchemaBuilder } from '../src/schema/schema-builder';
import {
  FedacoBuilderTestModelCloseRelatedStub, FedacoBuilderTestModelParentStub
} from './model/fedaco-builder-test-model-parent-stub';
import {
  FedacoBuilderTestModelSelfRelatedStub
} from './model/fedaco-builder-test-model-self-related-stub';

describe('fedaco builder relation', () => {

  class Conn implements ConnectionInterface {
    getQueryGrammar(): any {

    }

    getDatabaseName(): string {
      return 'default-database';
    }

    getPostProcessor(): any {

    }

    query(): QueryBuilder {
      return new QueryBuilder(
        this,
        new MysqlQueryGrammar(),
        new Processor()
      );
    }

    async select() {
      return await Promise.resolve();
    }

    async insert(sql: string, bindings: any[]): Promise<boolean> {
      throw new Error('not implement');
    }

    async update() {
    }

    async delete() {
    }

    async statement() {
    }

    async affectingStatement() {
    }

    getName() {
      return '';
    }

    getConfig(name: string): any {
    }

    getPdo(): any {
    }

    getSchemaBuilder(): SchemaBuilder {
      return undefined;
    }

    recordsHaveBeenModified(): any {
    }

    selectFromWriteConnection(sql: string, values: any): any {
    }

    table(table: Function | QueryBuilder | string, as?: string): QueryBuilder {
      return undefined;
    }

    insertGetId(sql: string, bindings: any[], sequence?: string): Promise<any> | boolean {
      return undefined;
    }

    _transactions: number;
    _transactionsManager: DatabaseTransactionsManager;

    afterCommit(callback: Function): Promise<void> {
      return Promise.resolve(undefined);
    }

    beginTransaction(): Promise<void> {
      return Promise.resolve(undefined);
    }

    commit(): Promise<void> {
      return Promise.resolve(undefined);
    }

    rollBack(toLevel?: number | null): Promise<void> {
      return Promise.resolve(undefined);
    }

    setTablePrefix(prefix: string): any {
    }

    setTransactionManager(manager: DatabaseTransactionsManager): this {
      return undefined;
    }

    transaction(callback: (...args: any[]) => Promise<any>, attempts?: number): Promise<any> {
      return Promise.resolve(undefined);
    }

    transactionLevel(): number {
      return 0;
    }

    unsetTransactionManager(): void {
    }
  }

  beforeAll(() => {
    const connection = new Conn();

    const resolver: ConnectionResolverInterface = {
      getDefaultConnection(): any {
      },
      setDefaultConnection(name: string): any {
      },
      connection() {
        return connection;
      }
    };
    (Model as typeof Model).setConnectionResolver(resolver);
  });

  it('test with count', () => {
    const model1 = new FedacoBuilderTestModelParentStub();

    const builder1 = model1.newQuery().withCount('foo');
    const result   = builder1.toSql();

    expect(result.result).toBe(
      'SELECT `fedaco_builder_test_model_parent_stubs`.*, (SELECT count(*) FROM `fedaco_builder_test_model_close_related_stubs` WHERE `fedaco_builder_test_model_parent_stubs`.`foo_id` = `fedaco_builder_test_model_close_related_stubs`.`id`) AS `foo_count` FROM `fedaco_builder_test_model_parent_stubs`');
  });

  it('test with count and select', () => {
    const model1 = new FedacoBuilderTestModelParentStub();

    const builder1 = model1.newQuery().select('id').withCount('foo');
    const result   = builder1.toSql();

    expect(result.result).toBe(
      'SELECT `id`, (SELECT count(*) FROM `fedaco_builder_test_model_close_related_stubs` WHERE `fedaco_builder_test_model_parent_stubs`.`foo_id` = `fedaco_builder_test_model_close_related_stubs`.`id`) AS `foo_count` FROM `fedaco_builder_test_model_parent_stubs`');
  });

  it('test with count and merged wheres', () => {
    const model   = new FedacoBuilderTestModelParentStub();
    const builder = model.newQuery().select('id').withCount({
      'activeFoo': (q: FedacoBuilder) => {
        q.where('bam', '>', 'qux');
      }
    });

    const result = builder.toSql();

    expect(result).toEqual({
      result  : 'SELECT `id`, (SELECT count(*) FROM `fedaco_builder_test_model_close_related_stubs` WHERE `fedaco_builder_test_model_parent_stubs`.`foo_id` = `fedaco_builder_test_model_close_related_stubs`.`id` AND `bam` > ? AND `active` = ?) AS `active_foo_count` FROM `fedaco_builder_test_model_parent_stubs`',
      bindings: ['qux', true]
    });
  });

  it('test with count and global scope', () => {
    const model = new FedacoBuilderTestModelParentStub();

    FedacoBuilderTestModelCloseRelatedStub.addGlobalScope('withCount', (q: FedacoBuilder) => {
      return q.addSelect('id');
    });
    const builder = model.newQuery().select('id').withCount(['foo']);

    FedacoBuilderTestModelCloseRelatedStub.addGlobalScope('withCount', (q: FedacoBuilder) => {
    });
    const result = builder.toSql();

    expect(result.result).toBe(
      'SELECT `id`, (SELECT count(*) FROM `fedaco_builder_test_model_close_related_stubs` WHERE `fedaco_builder_test_model_parent_stubs`.`foo_id` = `fedaco_builder_test_model_close_related_stubs`.`id`) AS `foo_count` FROM `fedaco_builder_test_model_parent_stubs`');

  });

  it('test with count and constraints and having', () => {
    const model   = new FedacoBuilderTestModelParentStub();
    // @ts-ignore
    const builder = model.newQuery().where('bar', 'baz');
    builder.withCount({
      'foo': (q: FedacoBuilder) => {
        q.where('bam', '>', 'qux');
      }
    }).having('foo_count', '>=', 1);

    const result = builder.toSql();


    expect(result.result).toBe(
      'SELECT `fedaco_builder_test_model_parent_stubs`.*, (SELECT count(*) FROM `fedaco_builder_test_model_close_related_stubs` WHERE `fedaco_builder_test_model_parent_stubs`.`foo_id` = `fedaco_builder_test_model_close_related_stubs`.`id` AND `bam` > ?) AS `foo_count` FROM `fedaco_builder_test_model_parent_stubs` WHERE `bar` = ? HAVING `foo_count` >= ?');

    expect(result.bindings).toEqual(['qux', 'baz', 1]);

  });

  it('test with count and rename', () => {
    const model   = new FedacoBuilderTestModelParentStub();
    const builder = model.newQuery().withCount('foo as foo_bar');

    const result = builder.toSql();

    expect(result.result).toBe(
      'SELECT `fedaco_builder_test_model_parent_stubs`.*, (SELECT count(*) FROM `fedaco_builder_test_model_close_related_stubs` WHERE `fedaco_builder_test_model_parent_stubs`.`foo_id` = `fedaco_builder_test_model_close_related_stubs`.`id`) AS `foo_bar` FROM `fedaco_builder_test_model_parent_stubs`');

  });

  it('test with count multiple and partial rename', () => {
    const model   = new FedacoBuilderTestModelParentStub();
    const builder = model.newQuery().withCount(['foo as foo_bar', 'foo']);

    const result = builder.toSql();

    expect(result.result).toBe(
      'SELECT `fedaco_builder_test_model_parent_stubs`.*, (SELECT count(*) FROM `fedaco_builder_test_model_close_related_stubs` WHERE `fedaco_builder_test_model_parent_stubs`.`foo_id` = `fedaco_builder_test_model_close_related_stubs`.`id`) AS `foo_bar`, (SELECT count(*) FROM `fedaco_builder_test_model_close_related_stubs` WHERE `fedaco_builder_test_model_parent_stubs`.`foo_id` = `fedaco_builder_test_model_close_related_stubs`.`id`) AS `foo_count` FROM `fedaco_builder_test_model_parent_stubs`');
  });
  it('test has with constraints and having in subquery', () => {
    const model   = new FedacoBuilderTestModelParentStub();
    const builder = model.newQuery().where('bar', 'baz');
    builder.whereHas('foo', (q: FedacoBuilder) => {
      q.having('bam', '>', 'qux');
    }).where('quux', 'quuux');

    const result = builder.toSql();

    expect(result.result).toBe(
      'SELECT * FROM `fedaco_builder_test_model_parent_stubs` WHERE `bar` = ? AND EXISTS (SELECT * FROM `fedaco_builder_test_model_close_related_stubs` WHERE `fedaco_builder_test_model_parent_stubs`.`foo_id` = `fedaco_builder_test_model_close_related_stubs`.`id` HAVING `bam` > ?) AND `quux` = ?');

    expect(result.bindings).toEqual(['baz', 'qux', 'quuux']);
  });

  it('test has with constraints with or where and having inSubquery', () => {
    const model   = new FedacoBuilderTestModelParentStub();
    const builder = model.newQuery().where('name', 'larry');
    builder.whereHas('address', (q: FedacoBuilder) => {
      q.where((q: FedacoBuilder) => {
        q.where('zipcode', '90210');
        q.orWhere('zipcode', '90220');
      });
      q.having('street', '=', 'fooside dr');
    }).where('age', 29);

    const result = builder.toSql();

    expect(result.result).toBe(
      'SELECT * FROM `fedaco_builder_test_model_parent_stubs` WHERE `name` = ? AND EXISTS (SELECT * FROM `fedaco_builder_test_model_close_related_stubs` WHERE `fedaco_builder_test_model_parent_stubs`.`foo_id` = `fedaco_builder_test_model_close_related_stubs`.`id` AND (`zipcode` = ? OR `zipcode` = ?) HAVING `street` = ?) AND `age` = ?');

    expect(result.bindings).toEqual(['larry', '90210', '90220', 'fooside dr', 29]);
  });

  it('test has with constraints and join and having in subquery', () => {
    const model   = new FedacoBuilderTestModelParentStub();
    const builder = model.newQuery().where('bar', 'baz');
    builder.whereHas('foo', (q: FedacoBuilder) => {
      q.join('quuuux', (j: JoinClauseBuilder) => {
        j.where('quuuuux', '=', 'quuuuuux');
      });
      q.having('bam', '>', 'qux');
    }).where('quux', 'quuux');

    const result = builder.toSql();

    expect(result.result).toBe(
      'SELECT * FROM `fedaco_builder_test_model_parent_stubs` WHERE `bar` = ? AND EXISTS (SELECT * FROM `fedaco_builder_test_model_close_related_stubs` INNER JOIN `quuuux` ON `quuuuux` = ? WHERE `fedaco_builder_test_model_parent_stubs`.`foo_id` = `fedaco_builder_test_model_close_related_stubs`.`id` HAVING `bam` > ?) AND `quux` = ?');

    expect(result.bindings).toEqual(['baz', 'quuuuuux', 'qux', 'quuux']);
  });

  it('test has with constraints and having in subquery with count', () => {
    const model   = new FedacoBuilderTestModelParentStub();
    const builder = model.newQuery().where('bar', 'baz');
    builder.whereHas('foo', (q: FedacoBuilder) => {
      q.having('bam', '>', 'qux');
    }, '>=', 2).where('quux', 'quuux');

    const result = builder.toSql();

    expect(result.result).toBe(
      'SELECT * FROM `fedaco_builder_test_model_parent_stubs` WHERE `bar` = ? AND (SELECT count(*) FROM `fedaco_builder_test_model_close_related_stubs` WHERE `fedaco_builder_test_model_parent_stubs`.`foo_id` = `fedaco_builder_test_model_close_related_stubs`.`id` HAVING `bam` > ?) >= 2 AND `quux` = ?');

    expect(result.bindings).toEqual(['baz', 'qux', 'quuux']);

  });

  it('test with count and constraints with binding in select sub', () => {
    const model   = new FedacoBuilderTestModelParentStub();
    const builder = model.newQuery();
    builder.withCount({
      'foo': (q: FedacoBuilder) => {
        q.selectSub(
          model.newQuery()
            .where('bam', '=', 3)
            .selectRaw('count(0)'),
          'bam_3_count'
        );
      }
    });

    const result = builder.toSql();

    expect(result.result).toBe(
      'SELECT `fedaco_builder_test_model_parent_stubs`.*, (SELECT count(*) FROM `fedaco_builder_test_model_close_related_stubs` WHERE `fedaco_builder_test_model_parent_stubs`.`foo_id` = `fedaco_builder_test_model_close_related_stubs`.`id`) AS `foo_count` FROM `fedaco_builder_test_model_parent_stubs`');

    expect(result.bindings).toEqual([]);
  });

  it('test has nested with constraints', () => {
    const model   = new FedacoBuilderTestModelParentStub();
    const builder = model.newQuery().whereHas('foo', (q: FedacoBuilder) => {
      q.whereHas('bar', (q: FedacoBuilder) => {
        q.where('baz', 'bam');
      });
    }).toSql();
    const result  = model.newQuery().whereHas('foo.bar', (q: FedacoBuilder) => {
      q.where('baz', 'bam');
    }).toSql();
    expect(builder.result).toBe(result.result);
  });

  it('test has nested', () => {
    const model   = new FedacoBuilderTestModelParentStub();
    const builder = model.newQuery().whereHas('foo', (q: FedacoBuilder) => {
      q.has('bar');
    });
    const result  = model.newQuery().has('foo.bar').toSql();
    expect(result).toEqual(builder.toSql());
  });
  it('test or has nested', () => {
    const model   = new FedacoBuilderTestModelParentStub();
    const builder = model.newQuery().whereHas('foo', (q: FedacoBuilder) => {
      q.has('bar');
    }).orWhereHas('foo', (q: FedacoBuilder) => {
      q.has('baz');
    });
    const result  = model.newQuery().has('foo.bar').orHas('foo.baz').toSql();

    expect(builder.toSql().result).toBe(
      'SELECT * FROM `fedaco_builder_test_model_parent_stubs` WHERE EXISTS (SELECT * FROM `fedaco_builder_test_model_close_related_stubs` WHERE `fedaco_builder_test_model_parent_stubs`.`foo_id` = `fedaco_builder_test_model_close_related_stubs`.`id` AND EXISTS (SELECT * FROM `fedaco_builder_test_model_far_related_stubs` WHERE `fedaco_builder_test_model_close_related_stubs`.`id` = `fedaco_builder_test_model_far_related_stubs`.`fedaco_builder_test_model_close_related_stubs_id`)) OR EXISTS (SELECT * FROM `fedaco_builder_test_model_close_related_stubs` WHERE `fedaco_builder_test_model_parent_stubs`.`foo_id` = `fedaco_builder_test_model_close_related_stubs`.`id` AND EXISTS (SELECT * FROM `fedaco_builder_test_model_far_related_stubs` WHERE `fedaco_builder_test_model_close_related_stubs`.`id` = `fedaco_builder_test_model_far_related_stubs`.`fedaco_builder_test_model_close_related_stubs_id`))'
    );
    expect(builder.toSql().result).toBe(result.result);
  });
  it('test self has nested', () => {
    const model        = new FedacoBuilderTestModelSelfRelatedStub();
    const nestedSql    = model.newQuery().whereHas('parentFoo', (q: FedacoBuilder) => {
      q.has('childFoo');
    }).toSql();
    const dotSql       = model.newQuery().has('parentFoo.childFoo').toSql();
    const alias        = 'self_alias_hash';
    const aliasRegex   = /\b(fedaco_reserved_\d)(\b|$)/ig;
    const nestedSqlStr = nestedSql.result.replace(aliasRegex, alias);
    const dotSqlStr    = dotSql.result.replace(aliasRegex, alias);
    expect(dotSqlStr).toBe(nestedSqlStr);
  });

  it('test self has nested uses alias', () => {
    const model      = new FedacoBuilderTestModelSelfRelatedStub();
    let sql          = model.newQuery().has('parentFoo.childFoo').toSql().result;
    const alias      = 'self_alias_hash';
    const aliasRegex = /\b(fedaco_reserved_\d)(\b|$)/ig;
    sql              = sql.replace(aliasRegex, alias);
    expect(sql).toContain('`self_alias_hash`.`id` = `self_related_stubs`.`parent_id`',);
  });

  it('test doesnt have', () => {
    const model   = new FedacoBuilderTestModelParentStub();
    const builder = model.newQuery().doesntHave('foo');

    const result = builder.toSql();

    expect(result.result).toBe(
      'SELECT * FROM `fedaco_builder_test_model_parent_stubs` WHERE NOT EXISTS (SELECT * FROM `fedaco_builder_test_model_close_related_stubs` WHERE `fedaco_builder_test_model_parent_stubs`.`foo_id` = `fedaco_builder_test_model_close_related_stubs`.`id`)');
  });

  it('test doesnt have nested', () => {
    const model   = new FedacoBuilderTestModelParentStub();
    const builder = model.newQuery().doesntHave('foo.bar');

    const result = builder.toSql();

    expect(result.result).toBe(
      'SELECT * FROM `fedaco_builder_test_model_parent_stubs` WHERE NOT EXISTS (SELECT * FROM `fedaco_builder_test_model_close_related_stubs` WHERE `fedaco_builder_test_model_parent_stubs`.`foo_id` = `fedaco_builder_test_model_close_related_stubs`.`id` AND EXISTS (SELECT * FROM `fedaco_builder_test_model_far_related_stubs` WHERE `fedaco_builder_test_model_close_related_stubs`.`id` = `fedaco_builder_test_model_far_related_stubs`.`fedaco_builder_test_model_close_related_stubs_id`))');
  });

  it('test or doesnt have', () => {
    const model   = new FedacoBuilderTestModelParentStub();
    const builder = model.newQuery().where('bar', 'baz').orDoesntHave('foo');

    const result = builder.toSql();

    expect(result).toEqual({
      result  : 'SELECT * FROM `fedaco_builder_test_model_parent_stubs` WHERE `bar` = ? OR NOT EXISTS (SELECT * FROM `fedaco_builder_test_model_close_related_stubs` WHERE `fedaco_builder_test_model_parent_stubs`.`foo_id` = `fedaco_builder_test_model_close_related_stubs`.`id`)',
      bindings: ['baz']
    });
  });

  it('test where doesnt have', () => {
    const model   = new FedacoBuilderTestModelParentStub();
    const builder = model.newQuery().whereDoesntHave('foo', (q: FedacoBuilder) => {
      q.where('bar', 'baz');
    });

    const result = builder.toSql();

    expect(result).toEqual(
      {
        result  : 'SELECT * FROM `fedaco_builder_test_model_parent_stubs` WHERE NOT EXISTS (SELECT * FROM `fedaco_builder_test_model_close_related_stubs` WHERE `fedaco_builder_test_model_parent_stubs`.`foo_id` = `fedaco_builder_test_model_close_related_stubs`.`id` AND `bar` = ?)',
        bindings: ['baz']
      });
  });

  it('test or where doesnt have', () => {
    const model   = new FedacoBuilderTestModelParentStub();
    const builder = model.newQuery().where('bar', 'baz').orWhereDoesntHave('foo',
      (q: FedacoBuilder) => {
        q.where('qux', 'quux');
      });

    const result = builder.toSql();

    expect(result).toEqual(
      {
        result  : 'SELECT * FROM `fedaco_builder_test_model_parent_stubs` WHERE `bar` = ? OR NOT EXISTS (SELECT * FROM `fedaco_builder_test_model_close_related_stubs` WHERE `fedaco_builder_test_model_parent_stubs`.`foo_id` = `fedaco_builder_test_model_close_related_stubs`.`id` AND `qux` = ?)',
        bindings: ['baz', 'quux']
      });
  });


});
