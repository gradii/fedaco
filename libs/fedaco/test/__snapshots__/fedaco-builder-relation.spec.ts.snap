// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fedaco builder relation test self has nested 1`] = `"SELECT * FROM \`self_related_stubs\` WHERE EXISTS (SELECT * FROM \`self_related_stubs\` AS \`self_alias_hash\` WHERE \`self_alias_hash\`.\`id\` = \`self_related_stubs\`.\`parent_id\` AND EXISTS (SELECT * FROM \`self_related_stubs\` WHERE \`self_alias_hash\`.\`id\` = \`self_related_stubs\`.\`parent_id\`))"`;

exports[`fedaco builder relation test self has nested uses alias 1`] = `"SELECT * FROM \`self_related_stubs\` WHERE EXISTS (SELECT * FROM \`self_related_stubs\` AS \`self_alias_hash\` WHERE \`self_alias_hash\`.\`id\` = \`self_related_stubs\`.\`parent_id\` AND EXISTS (SELECT * FROM \`self_related_stubs\` WHERE \`self_alias_hash\`.\`id\` = \`self_related_stubs\`.\`parent_id\`))"`;
